llama-index-core/llama_index/core/base/embeddings/base_sparse.py

```diff
@@ -0,0 +1,381 @@
+    if not embedding1 or not embedding2:
+        return 0.0
+
+    # Use the smaller embedding as the primary iteration set
+    if len(embedding1) > len(embedding2):
+        embedding1, embedding2 = embedding2, embedding1
+
+    # Precompute norms and find common indices
+    norm1 = norm2 = dot_product = 0.0
+    common_indices = set(embedding1.keys()) & set(embedding2.keys())
+
+    for idx, value in embedding1.items():
+        norm1 += value**2
+        if idx in common_indices:
+            dot_product += value * embedding2[idx]
+
+    for value in embedding2.values():
+        norm2 += value**2
+
+    if norm1 == 0.0 or norm2 == 0.0:
+        return 0.0
+
+    return dot_product / (math.sqrt(norm1) * math.sqrt(norm2))
+
+
+def mean_agg(embeddings: List[SparseEmbedding]) -> SparseEmbedding:
+    """Get mean aggregation of embeddings."""
+    if not embeddings:
+        return {}
+
+    sum_dict: Dict[int, float] = defaultdict(float)
+    for embedding in embeddings:
+        for idx, value in embedding.items():
+            sum_dict[idx] += value
+
+    return {idx: value / len(embeddings) for idx, value in sum_dict.items()}
+
+
+class BaseSparseEmbedding(BaseModel, DispatcherSpanMixin):
+    """Base class for embeddings."""
+
+    model_config = ConfigDict(
+        protected_namespaces=("pydantic_model_",), arbitrary_types_allowed=True
+    )
+    model_name: str = Field(
+        default="unknown", description="The name of the embedding model."
+    )
+    embed_batch_size: int = Field(
+        default=DEFAULT_EMBED_BATCH_SIZE,
+        description="The batch size for embedding calls.",
+        gt=0,
+        le=2048,
+    )
+    num_workers: Optional[int] = Field(
+        default=None,
+        description="The number of workers to use for async embedding calls.",
+    )
+
+    @abstractmethod
+    def _get_query_embedding(self, query: str) -> SparseEmbedding:
+        """
+        Embed the input query synchronously.
+
+        Subclasses should implement this method. Reference get_query_embedding's
+        docstring for more information.
+        """
+
+    @abstractmethod
+    async def _aget_query_embedding(self, query: str) -> SparseEmbedding:
+        """
+        Embed the input query asynchronously.
+
+        Subclasses should implement this method. Reference get_query_embedding's
+        docstring for more information.
+        """
+
+    @dispatcher.span
+    def get_query_embedding(self, query: str) -> SparseEmbedding:
+        """
+        Embed the input query.
+
+        When embedding a query, depending on the model, a special instruction
+        can be prepended to the raw query string. For example, "Represent the
+        question for retrieving supporting documents: ". If you're curious,
+        other examples of predefined instructions can be found in
+        embeddings/huggingface_utils.py.
+        """
+        model_dict = self.model_dump()
+        model_dict.pop("api_key", None)
```

Comment:
should `api_key` be an attribute? Or is this always going to be `None`?

Reply:
alternatively, i think there's a way for us to modify the serialization here to exclude `api_key` if there is one. If we add the method below then all `model_dump()` would include `class_name` and not have `api_key`

```python
from llama_index.core.bridge.pydantic import model_serializer

class MyBaseModel(BaseModel):

      @model_serializer(mode="wrap")
      def custom_model_dump(self, handler: Any) -> Dict[str, Any]:
          data = handler(self)
          # add class name
          data["class_name"] = self.class_name()
          # del api_key
          data.pop("api_key", None)
          return data
```

Reply:
Done!